### PR TITLE
changed dimensionality of Scores.scores

### PR DESCRIPTION
**Pull request by jspaaks**
This PR implements #72 

- similarity functions dimension is now always 1
- queries is now a vector
- added tests
- removed ``Scores.reverse`` method
- changed calculation of ``.sort_by()``, renamed to ``sort()`` and removed positional argument ``label`` 
- changed calculation of ``.top()``
- updated the readme figure to reflect these changes https://github.com/matchms/matchms/issues/78

### CODE REVIEW

File: integration-tests/test_user_workflow.py [hunk 1]
```diff
@@ -8,23 +8,28 @@
...
+    spectrums = [default_filters(s) for s in load_from_mgf(spectrums_file)]
 
-    query_spectrum = reference_spectrums[0].clone()
+    queries = spectrums[:7]
+    references = spectrums[6:]
 
-    # define similarity functions
-    similarity_functions = [IntersectMz("intersect")]
+    # define similarity function
+    intersect_mz = IntersectMz("intersect")
```
Comment (sverhoeven) [inline integration-tests/test_user_workflow.py]: Please remove label

File: integration-tests/test_user_workflow.py [hunk 2]
```diff
@@ -8,23 +8,28 @@
...
-    assert scores.scores[0] == 1., "Intersection of spectrum with itself should yield a perfect match."
-    assert scores.scores.shape == (10, 1), "Expected a table of 10 rows, 1 columns."
-    assert scores.scores[0] > scores.scores[1], "Expected a different sort order."
+    queries_top10, reference_top10, scores_top10, = scores.top(10, omit_self_comparisons=False)
+
+    print(scores_top10)
+
+    assert scores.scores[0, 6] == 1., "Intersection of spectrum with itself should yield a perfect match."
+    assert scores.scores.shape == (70, 7), "Expected a table of 10 rows, 1 columns."
+    assert queries_top10[0] == reference_top10[0], "Expected the best match between two copies of the same spectrum."
```
Comment (sverhoeven) [inline integration-tests/test_user_workflow.py:32]: == uses memory pointers to compare spectrums. Does the spectrum class have a __eq__ and partners https://www.python.org/dev/peps/pep-0207/?

File: matchms/Scores.py [hunk 3]
```diff
@@ -1,16 +1,16 @@
-import numpy
+from numpy import empty, argsort, unravel_index, hstack, vstack
 
 
 class Scores:
     """An example docstring for a class definition."""
-    def __init__(self, measured_spectrum, reference_spectrums, similarity_functions, scores=None):
+    def __init__(self, queries, references, similarity_function, scores=None):
```
Comment (fdiblen) [inline matchms/Scores.py]: scores should be private variable

File: matchms/Scores.py [hunk 4]
```diff
@@ -19,44 +19,28 @@ def __str__(self):
...
-                      reference_spectrums=reference_spectrums,
-                      similarity_functions=self.similarity_functions,
-                      scores=scores)
+    def top(self, n, kind="quicksort", omit_self_comparisons=True):
+
+        queries_sorted, references_sorted, scores_sorted = self.sort(kind=kind)
+
+        if omit_self_comparisons:
+            zipped = zip(queries_sorted, references_sorted, scores_sorted)
+            self_comparisons_omitted = [(q, r, s) for q, r, s in zipped if q != r]
```
Comment (sverhoeven) [inline matchms/Scores.py]: != is this the right comparison, if 2 spectrums have same peaks but different memory ref pointer then they are treated as different

File: integration-tests/test_user_workflow.py [hunk 5]
```diff
@@ -8,23 +8,28 @@
...
-                              reference_spectrums,
-                              similarity_functions).sort_by("intersect").reverse().top(10)
+    scores = calculate_scores(queries,
+                              references,
+                              intersect_mz)
 
-    assert scores.scores[0] == 1., "Intersection of spectrum with itself should yield a perfect match."
-    assert scores.scores.shape == (10, 1), "Expected a table of 10 rows, 1 columns."
-    assert scores.scores[0] > scores.scores[1], "Expected a different sort order."
+    queries_top10, reference_top10, scores_top10, = scores.top(10, omit_self_comparisons=False)
```
Comment (jspaaks) [inline integration-tests/test_user_workflow.py]: omit -> include, change default value accordingly

File: tests/test_scores.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    queries_top_2, references_top_2, scores_top_2 = scores.top(2, omit_self_comparisons=True)
+
+    assert numpy.all(queries_top_2 == numpy.asarray([["q1"], ["q0"]]))
+    assert numpy.all(references_top_2 == numpy.asarray([["r2"], ["q1"]]))
+    assert numpy.all(scores_top_2 == numpy.asarray([[6.], [5.]], dtype="float"))
+
+
+if __name__ == '__main__':
+    test_scores_init()
```
Comment (jspaaks) [inline tests/test_scores.py:45]: ~~replace with ``pytest.something``~~

File: matchms/Scores.py (hunk 4 above)
Comment (jspaaks) [reply matchms/Scores.py]: I'm trying to test for equality (``==`` and ``!=``), not identity (``is`` and ``not is``). The ``!=`` test should leave out a similarity score between equal spectrums, even when they are two copies of said spectrum, not two references to the same identity.

File: integration-tests/test_user_workflow.py (hunk 5 above)
Comment (jspaaks) [reply integration-tests/test_user_workflow.py]: changed

File: matchms/Scores.py (hunk 3 above)
Comment (jspaaks) [reply matchms/Scores.py]: changed

File: integration-tests/test_user_workflow.py (hunk 2 above)
Comment (jspaaks) [reply integration-tests/test_user_workflow.py:32]: added an ``__eq__`` with equality comparisons on ``mz``, ``intensities`` and ``metadata``.

File: integration-tests/test_user_workflow.py (hunk 1 above)
Comment (jspaaks) [reply integration-tests/test_user_workflow.py]: done

File: tests/test_scores.py (hunk 6 above)
Comment (jspaaks) [reply tests/test_scores.py:45]: this didnt work for me, so I'm leaving it for now.

File: matchms/Spectrum.py [hunk 7]
```diff
@@ -10,18 +10,21 @@ def __init__(self, mz, intensities, metadata=None):
...
         if metadata is None:
-            self._metadata = dict()
+            self.metadata = dict()
         else:
-            self._metadata = metadata
+            self.metadata = metadata
+
+    def __eq__(self, other):
+        return \
+            numpy.all(self.mz == other.mz) and \
```
Comment (sverhoeven) [inline matchms/Spectrum.py]: Will this not compare all digits of a float, causing too many non equality errors. Better to use [allclose](https://docs.scipy.org/doc/numpy/reference/generated/numpy.allclose.html#numpy.allclose)

Comment (jspaaks) [reply matchms/Spectrum.py]: good find, didnt know that one

**Review by sverhoeven — APPROVED** [review]

File: matchms/Scores.py (hunk 3 above)
Comment (jspaaks) [reply matchms/Scores.py]: @fdiblen can you mark this as resolved if you think it is, then approve the PR if you find no further issues with it? Thanks!

**Review by fdiblen — APPROVED** [review]

File: tests/test_scores.py (hunk 6 above)
Comment (jspaaks) [reply tests/test_scores.py:45]: Possibly found the reason https://github.com/pytest-dev/pytest-cov/issues/131 (coverage calculation)

**Pull request merged**